src/main/java/uk/gov/ons/census/casesvc/service/ReceiptService.java

```diff
@@ -20,41 +20,103 @@
     this.caseService = caseService;
     this.uacService = uacService;
     this.eventLogger = eventLogger;
+    this.fieldworkFollowupService = fieldworkFollowupService;
   }
 
   public void processReceipt(ResponseManagementEvent receiptEvent) {
     ResponseDTO receiptPayload = receiptEvent.getPayload().getResponse();
     UacQidLink uacQidLink = uacService.findByQid(receiptPayload.getQuestionnaireId());
+
+    logEvent(receiptEvent, receiptPayload, uacQidLink);
+
+    // If the uacQidLink is already marked as a blank questionnaire, the uacQidLink is inactive and
+    // can't be changed
+    if (uacQidLink.isBlankQuestionnaireReceived()) {
+      return;
+    }
+
+    if (receiptPayload.isUnreceipt()) {
+      processUnreceiptingEvent(uacQidLink, receiptEvent);
+      return;
+    }
+
+    if (isCCSQuestionnaireType(uacQidLink.getQid())) {
+      processCCSQid(uacQidLink, receiptEvent);
+      return;
+    }
+
+    processReceiptForStandardCase(uacQidLink, receiptEvent);
+  }
+
+  private void processReceiptForStandardCase(
+      UacQidLink uacQidLink, ResponseManagementEvent receiptEvent) {
     uacQidLink.setActive(false);
 
+    uacService.saveAndEmitUacUpdatedEvent(uacQidLink, false);
+
     Case caze = uacQidLink.getCaze();
 
     if (caze != null) {
       caze.setReceiptReceived(true);
+      caseService.saveAndEmitCaseUpdatedEvent(caze);
+    } else {
+      logUnlinkedUacQidLink(uacQidLink, receiptEvent);
+    }
+  }
 
-      if (caze.isCcsCase()) {
-        caseService.saveCase(caze);
-      } else {
-        caseService.saveAndEmitCaseUpdatedEvent(caze);
-      }
+  private void processCCSQid(UacQidLink uacQidLink, ResponseManagementEvent receiptEvent) {
+    uacQidLink.setActive(false);
+
+    uacService.saveUacQidLink(uacQidLink);
+
+    Case caze = uacQidLink.getCaze();
+
+    if (caze != null) {
+      // Not currently testing for horror of unreceipted CCS? is this possible?
+
+      caze.setReceiptReceived(true);
+      caseService.saveCase(caze);
     } else {
-      log.with("qid", receiptPayload.getQuestionnaireId())
-          .with("tx_id", receiptEvent.getEvent().getTransactionId())
-          .with("channel", receiptEvent.getEvent().getChannel())
-          .warn("Receipt received for unaddressed UAC/QID pair not yet linked to a case");
+      logUnlinkedUacQidLink(uacQidLink, receiptEvent);
     }
+  }
 
-    if (isCCSQuestionnaireType(uacQidLink.getQid())) {
-      uacService.saveUacQidLink(uacQidLink);
+  private void processUnreceiptingEvent(
+      UacQidLink uacQidLink, ResponseManagementEvent receiptEvent) {
+
+    uacQidLink.setBlankQuestionnaireReceived(true);
+    uacService.saveAndEmitUacUpdatedEvent(uacQidLink, true);
+
+    Case caze = uacQidLink.getCaze();
+
+    if (caze != null) {
+      if (hasCaseAlreadyBeenReceiptedByAnotherQid(uacQidLink)) {
+        return;
+      }
+
+      caze.setReceiptReceived(false);
+      fieldworkFollowupService.buildAndSendFieldWorkFollowUp(caze);
```

Comment:
oooh... maybe my earlier comments were flawed. I see where this is getting used other than CCS now